lib/hcrypto/evp-pkcs11.c

```diff
@@ -0,0 +1,744 @@
+/*
+ * Copyright (c) 2015, Secure Endpoints Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * - Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ *
+ * - Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in
+ *   the documentation and/or other materials provided with the
+ *   distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* PKCS#11 provider */
+
+#include "config.h"
+
+#include <baselocl.h>
+
+#include <assert.h>
+#ifdef HAVE_DLFCN_H
+#include <dlfcn.h>
+#ifndef RTLD_LAZY
+#define RTLD_LAZY 0
+#endif
+#ifndef RTLD_LOCAL
+#define RTLD_LOCAL 0
+#endif
+#ifndef RTLD_GROUP
+#define RTLD_GROUP 0
+#endif
+#else
+#error PKCS11 support requires dlfcn.h
+#endif
+
+#include <evp.h>
+#include <evp-hcrypto.h>
+#include <evp-pkcs11.h>
+
+#include "pkcs11.h"
+
+#if __sun && !defined(PKCS11_MODULE_PATH)
+# if _LP64
+# define PKCS11_MODULE_PATH "/usr/lib/64/libpkcs11.so"
+# else
+# define PKCS11_MODULE_PATH "/usr/lib/libpkcs11.so"
+# endif
+#endif
+
+static CK_FUNCTION_LIST_PTR p11_module;
+
+static int
+p11_cleanup(EVP_CIPHER_CTX *ctx);
+
+struct pkcs11_cipher_ctx {
+    CK_SESSION_HANDLE hSession;
+    CK_OBJECT_HANDLE hSecret;
+};
+
+struct pkcs11_md_ctx {
+    CK_SESSION_HANDLE hSession;
+};
+
+static CK_RV
+p11_module_init(void)
+{
+    CK_RV rv = CKR_OK;
+
+    if (p11_module == NULL) {
+        CK_RV (*C_GetFunctionList_fn)(CK_FUNCTION_LIST_PTR_PTR);
+        CK_FUNCTION_LIST_PTR module;
+        void *handle = NULL;
+
+        if (!issuid()) {
+            char *pkcs11ModulePath = getenv("PKCS11_MODULE_PATH");
+            if (pkcs11ModulePath != NULL) {
+                handle = dlopen(pkcs11ModulePath, RTLD_LAZY | RTLD_LOCAL | RTLD_GROUP);
+                if (handle == NULL)
+                    fprintf(stderr, "p11_module_init(%s): %s\n", pkcs11ModulePath, dlerror());
+            }
+        }
+#ifdef PKCS11_MODULE_PATH
```

Comment:
> On 29 Nov 2015, at 11:05 AM, Nico Williams <notifications@github.com> wrote:
> 
> Also, the "there's two paths, one for the system 32-bit PKCS11 provider, and the other for the system 64-bit PKCS11 provider" problem is likely to come up elsewhere. Maybe we need a lib/base utility for constructing dlopen()/LoadLibraryEx()/whatever paths that supports a template where whatever appropriate token for the bit-ness of the running process can be inserted.
> 
Is this a common issue though? On Solaris it seems to be a problem because uname/autoconf doesn’t give us a way to determine if we’re building 64-bit. On other platforms I think it’s a non-issue – OS X has universal binaries, and every other platform I can think of has 64-bit as a separate target.

— Luke